### ISSUE TITLE
pivottabler throws error if the names of columns in the data frame contains space

### ISSUE THREAD

**Issue by monicarajamanohar** [cbailiss/pivottabler#14]
Hi,

I have a data frame whose column names contains space. For example "AAs Desiganation" is one of the columns in my data frame. When I run the code, I receive the following error. 

Warning: Error in parse: <text>:1:31: unexpected symbol
1: df <- dplyr::distinct(df, AAs Desiganation
                                  ^
  76: parse
  74: private$p_columnGroup$addDataGroups
  73: pt$addColumnDataGroups

Thanks,
Monica.

**Comment by cbailiss** [comment]
Thanks for the bug report.  I'll look into it at some point over the next few days and post updates here.

**Comment by cbailiss** [comment]
I've found a few minutes to quickly look into this.  

This is an interesting point, because generally spaces are to be avoided in column/variable names in R as it makes life difficult, e.g. they aren't valid in identifiers.  Some references:
https://dnidzgorski.wordpress.com/2017/06/09/r-fix-column-names-spaces/
https://cran.r-project.org/doc/manuals/r-release/R-lang.html#Identifiers
https://www.rdocumentation.org/packages/base/versions/3.5.1/topics/data.frame
(by default when creating a data frame make.names() removes spaces).

That said, some commonly used R libraries do support spaces, so it is perhaps more common than I thought.  It would help me to understand (briefly) - why you need to work with column names containing spaces?  

For pivottabler, the column names aren't displayed in the pivot table and a caption can be specified for calculations (which can contain spaces).

Example:

```
# try to specify a data frame with spaces in names
mydata <- data.frame("Sale ID"=1:5, "Colour"=c("Red", "Red", "Green", "Red", "Green"), "Sale Item"=c("Car", "Lorry", "Car", "Train", "Train"), "Sale Amount"=c(12.1,2.3,5.6,3.7,1.5))

# the following returns:  "Sale.ID"  "Colour"  "Sale.Item"  "Sale.Amount"
# i.e. R has tried to dissuade the use of space in names to satisfy the requirements of identifiers
colnames(mydata)

# display a pivot table, specifying calculation names with spaces 
library(pivottabler)
pt <- PivotTable$new()
pt$addData(mydata)
pt$addColumnDataGroups("Sale.Item")
pt$addRowDataGroups("Colour")
pt$defineCalculation(calculationName="SaleCount", caption="Sale Count", summariseExpression="n()")
pt$defineCalculation(calculationName="TotalAmount", caption="Total Amount", summariseExpression="sum(Sale.Amount)")
pt$renderPivot()
```

**Footnote - reproducing the issue (or a similar issue):**

```
# try to specify a data frame with spaces in names
mydata <- data.frame("Sale ID"=1:5, "Colour"=c("Red", "Red", "Green", "Red", "Green"), "Sale Item"=c("Car", "Lorry", "Car", "Train", "Train"), "Sale Amount"=c(12.1,2.3,5.6,3.7,1.5))

# the following returns:  "Sale.ID"  "Colour"  "Sale.Item"  "Sale.Amount"
# i.e. R has tried to dissuade the use of space in names to satisfy the requirements of identifiers
colnames(mydata)

# forcing the names with spaces
colnames(mydata) <- c("Sale ID", "Colour", "Sale Item", "Sale Amount")

# the following returns:  "Sale ID"  "Colour"  "Sale Item"  "Sale Amount"
colnames(mydata)

# testing the pivot table
library(pivottabler)
pt <- PivotTable$new()
pt$addData(mydata)
pt$addColumnDataGroups("Sale Item")
pt$addRowDataGroups("Colour")
pt$defineCalculation(calculationName="Total Sales", summariseExpression="n()")
pt$renderPivot()
```

Fails with:
Error in parse(text = paste0("df <- dplyr::distinct(df, ", paste(varNamesInUse3,  : 
  <text>:1:32: unexpected symbol
1: df <- dplyr::distinct(df, Sale Item

**Comment by monicarajamanohar** [comment]
Thank you for spending so much time for my issue. In my case, I am writing a R shiny application which will provide the users with ability to read from "views" of a postgresql database and perform some manipulation/analysis such as pivoting. So I have no control over the columns or the names of columns being returned. Many of the column names contain spaces.

**Comment by cbailiss** [comment]
I have spent some time today improving support for unusual/illegal names in the pivottabler package.  The example code (in the footnote of my comment above), that failed before with an error, now runs without any error.

Please could you test if this works for and let me know if you have any problems.

The pivottabler changes are present in GitHub only.  The changes are not present in the release version (and won't be for a while), so you need to install the updated version of the package from GitHub.  To install the package from GitHub, install the devtools package (if you don't have it) and run the following:
devtools::install_github("cbailiss/pivottabler")

More information:

An example of what now works when referring to data frame columns:
`pt$addColumnDataGroups("Sale Item")`

Note - unusual/illegal column names (e.g. containing spaces) must be wrapped in back-ticks in summarise expressions and calculation expressions (but there is no need to do this for calculation names), e.g.
```
pt$defineCalculation(calculationName="Total Sales", summariseExpression="sum(`Sale Amount`)")
pt$defineCalculation(type="calculation", basedOn=c("Total Sales", "Sale Count"), format="%.1f", 
                     calculationName="Avg Sale Amount", calculationExpression="values$`Total Sales`/values$`Sale Count`")

```

**Comment by cbailiss** [comment]
Resolved in v1.0.0
